api/retry.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package api
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"time"
+)
+
+type client interface {
+	put(endpoint string, in, out any, q *WriteOptions) (*WriteMeta, error)
+}
+
+type retryClient struct {
+	c client
+	retryOptions
+}
+
+// LockOptions is used to parameterize the Lock behavior.
+type retryOptions struct {
+	MaxRetries      uint64        // Optional, defaults to 3
+	MaxBetweenCalls time.Duration // Optional, defaults to 0, meaning no time cap
+	MaxToLastCall   time.Duration // Optional, defaults to 0, meaning no time cap
+	FixedDelay      time.Duration // Optional, defaults to 0, meaning Delay is exponential, starting at 1sec
+	DelayBase       time.Duration // Optional, defaults to 1sec
+}
+
+func newRetryClient(c client, opts retryOptions) *retryClient {
+	rc := &retryClient{
+		c: c,
+		retryOptions: retryOptions{
+			MaxRetries: 3,
+			DelayBase:  time.Second,
+		},
+	}
+
+	if opts.DelayBase != 0 {
+		rc.DelayBase = opts.DelayBase
+	}
+
+	if opts.MaxRetries != 0 {
+		rc.MaxRetries = opts.MaxRetries
+	}
+
+	if opts.MaxBetweenCalls != 0 {
+		rc.MaxBetweenCalls = opts.MaxBetweenCalls
+	}
+
+	if opts.MaxToLastCall != 0 {
+		rc.MaxToLastCall = opts.MaxToLastCall
+	}
+
+	if opts.FixedDelay != 0 {
+		rc.FixedDelay = opts.FixedDelay
+	}
+
+	return rc
+}
+
+func (rc *retryClient) retryPut(ctx context.Context, endpoint string, in, out any, q *WriteOptions) (*WriteMeta, error) {
+	var err error
+	var wm *WriteMeta
+
+	iDelay := time.Duration(0)
```

Comment:
`iDelay` for initial delay? If so, we use it for all delays, so maybe `delayDuration`?

Reply:
it was for iteration delay, but lets change it for attempt delay :)